File: missing-ranges/missing-ranges.java
class Solution {
    public List<String> findMissingRanges(int[] nums, int lower, int upper) {
        List<String> ans = new ArrayList<>();
        int next = lower;
        for (int num : nums) {
            if (num > next) {
                addRange(ans, next, num - 1);
            }
            next = num + 1;
        }
        
        if (next <= upper) {
            addRange(ans, next, upper);
        }
        
        return ans;
    }
    
    private void addRange(List<String> list, int start, int end) {
        if (start == end) {
            list.add(String.valueOf(start));
        } else {
            list.add(String.format("%d->%d", start, end));
        }
    }
}

File: missing-ranges/missing-ranges.py
class Solution:
    def findMissingRanges(self, nums: List[int], lower: int, upper: int) -> List[str]:
        
        missing_ranges = []
        
        def add_range(start, end):
            if start > end:
                return
            
            if start == end:
                missing_ranges.append(str(start))
            else:
                missing_ranges.append(f'{start}->{end}')
        
        if not nums:
            add_range(lower, upper)
            return missing_ranges
        
        add_range(lower, nums[0] - 1)
        for i in range(1, len(nums)):
            prev = nums[i - 1]
            curr = nums[i]
            
            start = max(lower, prev + 1)
            end = min(upper, curr - 1)
            add_range(start, end)
        add_range(nums[-1] + 1, upper)
        
        return missing_ranges